python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,652 @@
+        handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+
+    vecs_cai = new_vectors.__cuda_array_interface__
+    vecs_dt = np.dtype(vecs_cai["typestr"])
+    cdef uint64_t n_rows = vecs_cai["shape"][0]
+    cdef uint32_t dim = vecs_cai["shape"][1]
+
+    _check_input_array(vecs_cai, [np.dtype('float32'), np.dtype('byte'), np.dtype('ubyte')], 
+                       exp_cols=index.dim)
+
+    idx_cai = new_indices.__cuda_array_interface__
+    _check_input_array(idx_cai, [np.dtype('uint64')], exp_rows=n_rows)
+    if len(idx_cai["shape"])!=1:
+        raise ValueError("Indices array is expected to be 1D")
+
+
+    cdef uintptr_t vecs_ptr = vecs_cai["data"][0]
+    cdef uintptr_t idx_ptr = idx_cai["data"][0]
+
+    if vecs_dt == np.float32:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <float*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)
+    elif vecs_dt == np.int8:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <int8_t*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)
+    elif vecs_dt == np.uint8:
+        c_ivf_pq.extend(deref(handle_),
+                        index.index,
+                        <uint8_t*>vecs_ptr,
+                        <uint64_t*> idx_ptr,
+                        <uint64_t> n_rows)       
+    else:
+        raise TypeError("query dtype %s not supported" % vecs_dt)
+
+    return index 
+
+
+cdef class SearchParams:
+    cdef c_ivf_pq.search_params params
+
+    def __init__(self, *, n_probes=20, 
+                 lut_dtype=CUDA_R_32F, 
+                 internal_distance_dtype=CUDA_R_32F):
+        """
+        IVF-PQ search parameters
+
+        Parameters
+        ----------
+        n_probes: int, default = 1024
+            The number of course clusters to select for the fine search.
+        lut_dtype: default = ivf_pq.CUDA_R_32F (float)
+            Data type of look up table to be created dynamically at search time. The use of 
+            low-precision types reduces the amount of shared memory required at search time, so
+            fast shared memory kernels can be used even for datasets with large dimansionality.
+            Note that the recall is slightly degraded when low-precision type is selected.
+            Possible values [CUDA_R_32F, CUDA_R_16F, CUDA_R_8U]
+        internal_distance_dtype: default = ivf_q.CUDA_R_32F (float)
+            Storage data type for distance/similarity computation.
+            Possible values [CUDA_R_32F, CUDA_R_16F]
+        
+        """
+
+        self.params.n_probes = n_probes
+        self.params.lut_dtype = lut_dtype
+        self.params.internal_distance_dtype = internal_distance_dtype
+        # self.params.shmem_carveout = self.shmem_carveout # TODO(tfeher): enable if #926 adds this
```

Comment:
We are going to want to consider eventually hiding this somehow behind more environment-specific set of parameters. This is leaking cuda-specific stuff into the public APIs (not something I think we need to change in this PR, but something we should be thinking about with Grace coming, and potentially having to support CPU code paths in the future)